src/metaprob/autodiff.cljc

```diff
@@ -0,0 +1,247 @@
+                        (new-f (:value x1) x2 x3)
+                        (* (df-dx1 (:value x1) x2 x3) (:derivative x1)))
+
+      (and (map? x2)
+           (or (non-dual? x1) (clojure.core/< (:tag x1) (:tag x2)))
+           (or (non-dual? x3) (clojure.core/< (:tag x3) (:tag x2))))
+      (make-dual-number (:tag x2)
+                        (new-f x1 (:value x2) x3)
+                        (* (df-dx2 x1 (:value x2) x3) (:derivative x2)))
+
+      (and (map? x3)
+           (or (non-dual? x1) (clojure.core/< (:tag x1) (:tag x3)))
+           (or (non-dual? x2) (clojure.core/< (:tag x2) (:tag x3))))
+      (make-dual-number (:tag x3)
+                        (new-f x1 x2 (:value x3))
+                        (* (df-dx3 x1 x2 (:value x3)) (:derivative x3)))
+
+      (and (map? x1) (map? x2) (= (:tag x1) (:tag x2)) (or (non-dual? x3) (clojure.core/< (:tag x3) (:tag x1))))
+      (make-dual-number (:tag x1)
+                        (new-f (:value x1) (:value x2) x3)
+                        (+ (* (df-dx1 (:value x1) (:value x2) x3) (:derivative x1))
+                           (* (df-dx2 (:value x1) (:value x2) x3) (:derivative x2))))
+
+      (and (map? x1) (map? x3) (= (:tag x1) (:tag x3)) (or (non-dual? x2) (clojure.core/< (:tag x2) (:tag x1))))
+      (make-dual-number (:tag x1)
+                        (new-f (:value x1) x2 (:value x3))
+                        (+ (* (df-dx1 (:value x1) x2 (:value x3)) (:derivative x1))
+                           (* (df-dx3 (:value x1) x2 (:value x3)) (:derivative x3))))
+
+      (and (map? x2) (map? x3) (= (:tag x2) (:tag x3)) (or (non-dual? x1) (clojure.core/< (:tag x1) (:tag x2))))
+      (make-dual-number (:tag x2)
+                        (new-f x1 (:value x2) (:value x3))
+                        (+ (* (df-dx2 x1 (:value x2) (:value x3)) (:derivative x2))
+                           (* (df-dx3 x1 (:value x2) (:value x3)) (:derivative x3))))
+
+      (and (map? x1) (map? x2) (map? x3) (= (:tag x1) (:tag x2) (:tag x3)))
+      (make-dual-number (:tag x2)
+                        (new-f (:value x1) (:value x2) (:value x3))
+                        (+ (* (df-dx1 (:value x1) (:value x2) (:value x3)) (:derivative x1))
+                           (* (df-dx2 (:value x1) (:value x2) (:value x3)) (:derivative x2))
+                           (* (df-dx3 (:value x1) (:value x2) (:value x3)) (:derivative x3)))))))
+
+;; For functions like + and *, which can take multiple arguments
+(defn lift-real-n->real [f df-dx1 df-dx2]
+  (fn [& xs]
+    (if (nil? xs)
+      (f)
+      (reduce (lift-real*real->real f df-dx1 df-dx2) xs))))
+
+;; For functions like - and /, where the first argument is special
+(defn lift-real-n+1->real [f df-dx df-dx1 df-dx2]
+  (fn [& xs]
+    (cond (empty? xs) (f)
+          (empty? (rest xs)) ((lift-real->real f df-dx) (first xs))
+          :else (reduce (lift-real*real->real f df-dx1 df-dx2) xs))))
+
+(defn value [x]
+  (if (non-dual? x) x (:value x)))
+
+
+(defn lift-real-n->boolean [f]
```

Comment:
Does this effectively mean "give up on differentiating if you encounter a boolean-valued function"?

Reply:
Not quite -- we can still do AD through functions that use operations like `<`.

For example, if `x` is a dual number:
```
(if (f x) (g x) (h x))
```
will also be one, so long as `g` and `h` are "lifted" `real->real` functions. Just by the way that Clojure works, nothing that `f`  does could stop that from being the case.

If `f` is a function like `(fn [x] (and (< x 3) (> x 0)))`, it will work fine in the above code without "lifting" (so long as `<` and `>` are lifted). The point of the `lift` operator is to take a primitive function that depends on its argument being (at the bits level) a number (like `<`, which ends up invoking some CPU instruction on numbers), and makes it work with dual-number arguments too, by extracting the `:value` field first.

Reply:
I see -- so there is an assumption that the boolean-valued function is piecewise continuous (i.e. piecewise constant).  Thus the derivative of `(if (f x) (g x) (h x))` is `(if (f x) (Dg x) (Dh x))`.  Is that right?

Reply:
Yes, exactly. This is the case for all the Boolean functions we lift.